vermouth/map_input.py

```diff
@@ -182,3 +186,104 @@ def read_mapping_directory(directory):
+        themselves as values. The blocks must be instances of :class:`nx.Graph`
+        with each node having an 'atomname' attribute.
+
+    Returns
+    -------
+    mappings: dict
+        A dictionary of mappings where the keys are the names of the blocks,
+        and the values are tuples like (mapping, weights, extra). The elements
+        of these tuples are formatted as the corresponding output of the
+        :fun:`read_mapping` function.
+
+    Raises
+    ------
+    KeyError
+        Raised if a node does not have am 'atomname' attribute.
+
+    See Also
+    --------
+    read_mapping
+        Read a mapping from a file.
+    generate_all_self_mappings
+        Generate self mappings for a list of force fields.
+    """
+    mappings = {}
+    for name, block in blocks.items():
+        mapping = {
+            (0, atom['atomname']): [(0, atom['atomname'])]
+            for atom in block.nodes.values()
+        }
+        weights = {
+            (0, atom['atomname']): {(0, atom['atomname']): 1}
+            for atom in block.nodes.values()
+        }
+        extra = []
+        mappings[name] = (mapping, weights, extra)
+    return mappings
+
+
+def generate_all_self_mappings(force_fields):
+    """
+    Generate self mappings for a list of force fields.
+
+    Parameters
+    ----------
+    force_fields: Iterable
+        List of instances of :class:`ForceField`.
+
+    Returns
+    -------
+    dict
+        A collection of mappings formatted as the output of the
+        :fun:`read_mapping_directory` function.
+    """
+    mappings = collections.defaultdict(dict)
+    for name, force_field in force_fields.items():
+        mappings[name][name] = generate_self_mappings(force_field.blocks)
+    return mappings
+
+
+def combine_mappings(known_mappings, partial_mapping):
+    """
+    Update a collection of mappings.
+
+    Add the mappings from the 'partial_mapping' argument into the
+    'known_mappings' collection. Both arguments are collections of mappings
+    similar to the output of the :fun:`read_mapping_directory` function. They
+    are dictionary with 3 levels of keys: the name of the initial force field,
+    the name of the target force field, and the name of the block. The values
+    in the third level dictionary are tuples of (mapping, weights, extra) as
+    described in the :fun:`read_mapping`.
+
+    If a force field appears in 'partial_mapping' that is not in
+    'known_mappings', then it is added. For existing pairs of initial and
+    target force fields, the blocks are updated and the version in
+    'partial_mapping' is kept in priority.
+
+    Parameters
+    ----------
+    known_mappings: dict
+        Collection of mapping to update **in-place**.
+    partial_mapping: dict
+        Collection of mappings to update from.
+    """
+    for origin, destinations in partial_mapping.items():
+        known_mappings[origin] = known_mappings.get(origin, {})
```

Comment:
This looks like a defaultdict of a defaultdict again, but I'm not sure we can use that here since it's an existing datastructure. I guess we can't :(